### PR TITLE
growing bucket optimization for greedy text generation

### PR DESCRIPTION
**Pull request by ssarkar2**
# What does this PR do?
### TL;DR
Instead of padding up the kvcache up to full size before starting, we grow the cache/input in multiples of bucketsize

### Details
let `p` be prompt len, `m` be max_new_tokens, `b` be bucketsize
We start with `shape = b*ceil(p/b)` (the [smallest multiple of b larger than p](https://github.com/huggingface/optimum-habana/blob/sarkar/bucket_text_gen/optimum/habana/transformers/generation/utils.py#L1246))
Once we fill it up, we move to next shape by adding `b` [more space](https://github.com/huggingface/optimum-habana/blob/sarkar/bucket_text_gen/optimum/habana/transformers/generation/utils.py#L1253) (`shape = shape + b`), and [copying](https://github.com/huggingface/optimum-habana/blob/sarkar/bucket_text_gen/optimum/habana/transformers/generation/utils.py#L1308) the old contents into the new expanded space

### Advantages
If we have multiple sentences of different lengths p1,p2,p3, then each time we run, we will compile for shapes p1+m, p2+m, p3+m if we pad to max
With this bucketing we can reuse old shapes
Also pad to max wastes a lot of computation to padding especially in the earlier runs of the generation. with bucketing, the initial smaller buckets can run faster.

### Test plan
Models: llama, falcon, bloom, gpt2
accuracy: fp32 test on piqa dataset, match with and without bucketing
performance: generate 256 tokens for 512, 1k, 2k prompt lens for llama
and some other tests

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by puneeshkhanna** [comment]
Have also verified this with HPU graphs enabled and also with limit hpu graphs ? Do you hit any memory limits and recompilations ?

**Comment by ssarkar2** [comment]
> Have also verified this with HPU graphs enabled and also with limit hpu graphs ? Do you hit any memory limits and recompilations ?

Yes, it has been tested in conjunction with limit_hpu.

File: optimum/habana/transformers/generation/configuration_utils.py [hunk 1]
```diff
@@ -31,5 +31,6 @@ def __init__(self, **kwargs):
         self.static_shapes = kwargs.get("static_shapes", None)
         self.ignore_eos = kwargs.get("ignore_eos", None)
         self.attn_softmax_bf16 = kwargs.get("attn_softmax_bf16", None)
+        self.bucketsize = kwargs.get("bucketsize", -1)
```
Comment (regisss) [inline optimum/habana/transformers/generation/configuration_utils.py:34]: Could you add it to the docstring above please?

File: examples/text-generation/README.md [hunk 2]
```diff
@@ -149,3 +149,18 @@ python run_generation.py \
 --prompt "Here is my prompt" \
 --peft_model trl-lib/llama-7b-se-rm-peft
 ```
+
+### Using growing bucket optimization
+With `--bucketsize`, instead of padding up the kvcache up to full size before starting, we grow the cache/input in multiples of bucketsize. This helps increase throughput and also reduce number of compilations if teh dataset has varying prompt lengths.
```
Comment (regisss) [inline examples/text-generation/README.md:154]:
```suggestion
With `--bucket_size`, instead of padding up the kv-cache up to full size before starting, we grow the cache/input in multiples of `bucket_size`. This helps increase throughput and also reduce number of compilations if the dataset has varying prompt lengths.
```

File: examples/text-generation/README.md [hunk 3]
```diff
@@ -149,3 +149,18 @@ python run_generation.py \
...
+
+For now, it is available only for greedy generation, and cannot be used with `--reuse_cache`
```
Comment (regisss) [inline examples/text-generation/README.md:156]:
```suggestion
> For now, it is available only for greedy generation, and cannot be used with `--reuse_cache`.
```

File: examples/text-generation/README.md [hunk 4]
```diff
@@ -149,3 +149,18 @@ python run_generation.py \
...
+```
```
Comment (regisss) [inline examples/text-generation/README.md:157]:
```suggestion

Here is an example:
```
```

File: examples/text-generation/README.md [hunk 5]
```diff
@@ -149,3 +149,18 @@ python run_generation.py \
...
+python run_generation.py \
+--model_name_or_path path_to_llama_or_falcon  \
+--use_hpu_graphs \
+--use_kv_cache \
+--bf16 \
+--max_new_tokens 200 \
+--batch_size=2 \
+--bucketsize 50
```
Comment (regisss) [inline examples/text-generation/README.md:165]:
```suggestion
--bucket_size 50
```

File: examples/text-generation/README.md [hunk 6]
```diff
@@ -149,3 +149,18 @@ python run_generation.py \
...
 --peft_model trl-lib/llama-7b-se-rm-peft
 ```
+
+### Using growing bucket optimization
+With `--bucketsize`, instead of padding up the kvcache up to full size before starting, we grow the cache/input in multiples of bucketsize. This helps increase throughput and also reduce number of compilations if teh dataset has varying prompt lengths.
+
+For now, it is available only for greedy generation, and cannot be used with `--reuse_cache`
+```
+python run_generation.py \
+--model_name_or_path path_to_llama_or_falcon  \
```
Comment (regisss) [inline examples/text-generation/README.md:159]:
```suggestion
--model_name_or_path path_to_model  \
```
It should be compatible with any model right?

File: optimum/habana/transformers/generation/configuration_utils.py (hunk 1 above)
Comment (regisss) [inline optimum/habana/transformers/generation/configuration_utils.py:34]:
```suggestion
        self.bucket_size = kwargs.get("bucket_size", -1)
```
Let's call it `bucket_size` to be consistent with `batch_size` and others.

File: examples/text-generation/run_generation.py [hunk 7]
```diff
@@ -177,6 +177,26 @@ def main():
         type=str,
         help="Output directory to store results in.",
     )
+    parser.add_argument(
+        "--bucketsize",
```
Comment (regisss) [inline examples/text-generation/run_generation.py:181]:
```suggestion
        "--bucket_size",
```

File: examples/text-generation/run_generation.py [hunk 8]
```diff
@@ -177,6 +177,26 @@ def main():
...
+        default=-1,
+        type=int,
+        help="Bucket size to maintain static shapes. If this number is negative (default is -1) \
+            then we use shape = prompt_length + max_new_tokens. If a positive number is passed \
```
Comment (regisss) [inline examples/text-generation/run_generation.py:185]:
```suggestion
            then we use `shape = prompt_length + max_new_tokens`. If a positive number is passed \
```

File: examples/text-generation/run_generation.py [hunk 9]
```diff
@@ -177,6 +177,26 @@ def main():
...
+            we increate the bucket in steps of bucketsize instead of allocating to max (prompt_length + max_new_tokens)",
```
Comment (regisss) [inline examples/text-generation/run_generation.py:186]:
```suggestion
            we increase the bucket in steps of `bucket_size` instead of allocating to max (`prompt_length + max_new_tokens`)",
```

File: examples/text-generation/run_generation.py [hunk 10]
```diff
@@ -371,6 +391,7 @@ def check_optimum_habana_min_version(*a, **b):
     generation_config.max_new_tokens = args.max_new_tokens
     generation_config.use_cache = args.use_kv_cache
     generation_config.static_shapes = is_optimized
+    generation_config.bucketsize = args.bucketsize if is_optimized else -1
```
Comment (regisss) [inline examples/text-generation/run_generation.py:394]:
```suggestion
    generation_config.bucket_size = args.bucket_size if is_optimized else -1
```

File: examples/text-generation/run_generation.py [hunk 11]
```diff
@@ -177,6 +177,26 @@ def main():
...
+        help="Bucket size to maintain static shapes. If this number is negative (default is -1) \
+            then we use shape = prompt_length + max_new_tokens. If a positive number is passed \
+            we increate the bucket in steps of bucketsize instead of allocating to max (prompt_length + max_new_tokens)",
+    )
+    parser.add_argument(
+        "--dataset_prompt_length",
+        default=16,
+        type=int,
+        help="Chop dataset to this length. If dataset_prompt_length<=0, do not chop (use full prompt). Default=16"
+    )
```
Comment (regisss) [inline examples/text-generation/run_generation.py:193]: I think it's similar to this PR that just got merged right?
https://github.com/huggingface/optimum-habana/pull/426

File: optimum/habana/transformers/generation/utils.py [hunk 12]
```diff
@@ -487,20 +487,29 @@ def generate(
                 inputs_tensor, generation_config.pad_token_id, generation_config.eos_token_id
             )
 
+        is_greedy_and_bucket = generation_config.bucketsize > 0 and self._get_generation_mode(generation_config, assistant_model) == GenerationMode.GREEDY_SEARCH
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:490]:
```suggestion
        is_greedy_and_bucket = generation_config.bucket_size > 0 and self._get_generation_mode(generation_config, assistant_model) == GenerationMode.GREEDY_SEARCH
```

File: optimum/habana/transformers/generation/utils.py [hunk 13]
```diff
@@ -487,20 +487,29 @@ def generate(
...
+        model_kwargs['bucketsize'] = generation_config.bucketsize if generation_config.static_shapes else -1
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:491]:
```suggestion
        model_kwargs['bucket_size'] = generation_config.bucket_size if generation_config.static_shapes else -1
```

File: optimum/habana/transformers/generation/utils.py [hunk 14]
```diff
@@ -487,20 +487,29 @@ def generate(
...
+        if generation_config.reuse_cache:
+            assert generation_config.bucketsize <= 0, "reuse_cache and bucketing flags set together"
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:493]:
```suggestion
            assert generation_config.bucket_size <= 0, "reuse_cache and bucketing flags set together"
```

File: optimum/habana/transformers/generation/utils.py [hunk 15]
```diff
@@ -487,20 +487,29 @@ def generate(
...
+                    model_kwargs["token_idx"] = torch.tensor(inputs_tensor.shape[-1], device=inputs_tensor.device)
+                    inputs_tensor = torch.nn.functional.pad(
+                        inputs_tensor, (0, generation_config.max_new_tokens), value=generation_config.pad_token_id
                     )
+                    if model_kwargs["attention_mask"] is not None:
+                        model_kwargs["attention_mask"] = torch.nn.functional.pad(
+                            model_kwargs["attention_mask"], (0, generation_config.max_new_tokens), value=0
+                        )
             else:
+                assert generation_config.bucketsize <= 0, "Untested path for bucket>0"
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:512]:
```suggestion
                assert generation_config.bucket_size <= 0, "Untested path for bucket>0"
```

File: optimum/habana/transformers/generation/utils.py [hunk 16]
```diff
@@ -487,20 +487,29 @@ def generate(
...
             if not self.config.is_encoder_decoder:
-                # token_idx is the current index in the generation process, it is incremented each time a new token is generated
-                model_kwargs["token_idx"] = torch.tensor(inputs_tensor.shape[-1], device=inputs_tensor.device)
-                inputs_tensor = torch.nn.functional.pad(
-                    inputs_tensor, (0, generation_config.max_new_tokens), value=generation_config.pad_token_id
-                )
-                if model_kwargs["attention_mask"] is not None:
-                    model_kwargs["attention_mask"] = torch.nn.functional.pad(
-                        model_kwargs["attention_mask"], (0, generation_config.max_new_tokens), value=0
+                # only pad if bucketsize < -1. If we are bucketing (bucketsize > 0), then that is taken care in greedy_search()
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:500]:
```suggestion
                # only pad if bucket_size < -1. If we are bucketing (bucket_size > 0), then that is taken care in greedy_search()
```

File: optimum/habana/transformers/generation/utils.py [hunk 17]
```diff
@@ -576,10 +585,17 @@ def generate(
                 calculated_max_length = input_ids.shape[-1] + generation_config.max_new_tokens
             if generation_config.use_cache and generation_config.reuse_cache:
                 bs, _ = input_ids.shape
-                unwrap_deepspeed_model(self).allocate_kv_cache(bs * generation_config.num_beams, calculated_max_length)
+                if not is_greedy_and_bucket:
+                    unwrap_deepspeed_model(self).allocate_kv_cache(bs * generation_config.num_beams, calculated_max_length)
 
         # 7. determine generation mode
         generation_mode = self._get_generation_mode(generation_config, assistant_model)
+        if generation_config.bucketsize > 0:
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:593]:
```suggestion
        if generation_config.bucket_size > 0:
```

File: optimum/habana/transformers/generation/utils.py [hunk 18]
```diff
@@ -576,10 +585,17 @@ def generate(
...
+            assert generation_config.static_shapes, "bucketsize > 0 can be set only when static_shapes is set"
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:594]:
```suggestion
            assert generation_config.static_shapes, "bucket_size > 0 can be set only when static_shapes is set"
```

File: optimum/habana/transformers/generation/utils.py [hunk 19]
```diff
@@ -576,10 +585,17 @@ def generate(
...
+        # if generation_config.bucketsize <= 0, padding is handled by the generating fn (like greedy_search)
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:595]:
```suggestion
        # if generation_config.bucket_size <= 0, padding is handled by the generating fn (like greedy_search)
```

File: optimum/habana/transformers/generation/utils.py [hunk 20]
```diff
@@ -576,10 +585,17 @@ def generate(
...
+        if generation_config.static_shapes and generation_config.bucketsize > 0:
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:596]:
```suggestion
        if generation_config.static_shapes and generation_config.bucket_size > 0:
```

File: optimum/habana/transformers/generation/utils.py [hunk 21]
```diff
@@ -576,10 +585,17 @@ def generate(
...
+            assert generation_mode == GenerationMode.GREEDY_SEARCH, "generation_config.bucketsize > 0 supported only for greedy mode"
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:597]:
```suggestion
            assert generation_mode == GenerationMode.GREEDY_SEARCH, "generation_config.bucket_size > 0 supported only for greedy mode"
```

File: optimum/habana/transformers/generation/utils.py [hunk 22]
```diff
@@ -623,6 +639,12 @@ def generate(
                 raise ValueError(
                     "You need to set `max_new_tokens` in your generation configuration to use static shapes."
                 )
+
+        if generation_config.static_shapes and generation_config.bucketsize > 0:
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:643]:
```suggestion
        if generation_config.static_shapes and generation_config.bucket_size > 0:
```

File: optimum/habana/transformers/generation/utils.py [hunk 23]
```diff
@@ -1211,6 +1233,29 @@ def greedy_search(
         hb_profer = HabanaProfile(warmup=profiling_warmup_steps, active=profiling_steps)
         hb_profer.start()
         this_peer_finished = False  # used by synced_gpus only
+        bucketsize = model_kwargs['bucketsize']
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1236]:
```suggestion
        bucket_size = model_kwargs['bucket_size']
```

File: optimum/habana/transformers/generation/utils.py [hunk 24]
```diff
@@ -1211,6 +1233,29 @@ def greedy_search(
...
+        def incrementor(bucketsize, prompt_len):
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1237]:
```suggestion
        def incrementor(bucket_size, prompt_len):
```

File: optimum/habana/transformers/generation/utils.py [hunk 25]
```diff
@@ -1211,6 +1233,29 @@ def greedy_search(
...
+            assert bucketsize > 0
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1238]:
```suggestion
            assert bucket_size > 0
```

File: optimum/habana/transformers/generation/utils.py [hunk 26]
```diff
@@ -1211,6 +1233,29 @@ def greedy_search(
...
+                else:
+                    token_idx += 1
+                    need_expansion = token_idx >= allocated_space
+                    if need_expansion:
+                        assert (allocated_space - token_idx) <= bucketsize
+                        allocated_space += bucketsize
+                yield {'allocated_space':allocated_space,'passnum':passnum,'token_idx':token_idx,'need_expansion':need_expansion}
+
+        prompt_len = input_ids.shape[-1]
+        if bucketsize >= 0:
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1255]:
```suggestion
        if bucket_size >= 0:
```

File: optimum/habana/transformers/generation/utils.py [hunk 27]
```diff
@@ -1211,6 +1233,29 @@ def greedy_search(
...
+            inc = iter(incrementor(bucketsize, prompt_len))
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1256]:
```suggestion
            inc = iter(incrementor(bucket_size, prompt_len))
```

File: optimum/habana/transformers/generation/utils.py [hunk 28]
```diff
@@ -1211,6 +1233,29 @@ def greedy_search(
...
+        if bucketsize > 0:
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1257]:
```suggestion
        if bucket_size > 0:
```

File: optimum/habana/transformers/generation/utils.py [hunk 29]
```diff
@@ -1211,6 +1233,29 @@ def greedy_search(
...
+                passnum+=1
+                if passnum == 0:
+                    token_idx = prompt_len
+                    allocated_space = int(math.ceil(prompt_len / bucketsize) * bucketsize)
+                    need_expansion = not (prompt_len == allocated_space)
+                else:
+                    token_idx += 1
+                    need_expansion = token_idx >= allocated_space
+                    if need_expansion:
+                        assert (allocated_space - token_idx) <= bucketsize
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1250]:
```suggestion
                        assert (allocated_space - token_idx) <= bucket_size
```

File: optimum/habana/transformers/generation/utils.py [hunk 30]
```diff
@@ -1211,6 +1233,29 @@ def greedy_search(
...
+                        allocated_space += bucketsize
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1251]:
```suggestion
                        allocated_space += bucket_size
```

File: optimum/habana/transformers/generation/utils.py [hunk 31]
```diff
@@ -1226,6 +1271,51 @@ def greedy_search(
                 if this_peer_finished_flag.item() == 0.0:
                     break
 
+            if bucketsize > 0:
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1274]:
```suggestion
            if bucket_size > 0:
```

File: optimum/habana/transformers/generation/utils.py [hunk 32]
```diff
@@ -1226,6 +1271,51 @@ def greedy_search(
...
+                # it will not have been padded if bucketsize > 0
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1275]:
```suggestion
                # it will not have been padded if bucket_size > 0
```

File: optimum/habana/transformers/generation/utils.py [hunk 33]
```diff
@@ -1226,6 +1271,51 @@ def greedy_search(
...
+                        model_kwargs["attention_mask"] = torch.nn.functional.pad(
+                            model_kwargs["attention_mask"], (0, pad_amount), value=0
+                        )
+                    else:
+                        assert False, "Not tested for cases where attn_mask isnt passed"
+
+                    if 'past_key_values' in model_kwargs:
+                        def create_pad_arg(pad_amount, i, j):
+                            if model_kwargs['past_key_values'][0][0].dim() == 3:
+                                assert 'GaudiBloomForCausalLM' in  str(type(self))  # bloom
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1294]:
```suggestion
                                assert self.config.model_type == "bloom"
```

File: optimum/habana/transformers/generation/utils.py [hunk 34]
```diff
@@ -1226,6 +1271,51 @@ def greedy_search(
...
+                            elif model_kwargs['past_key_values'][0][0].dim() == 4:
+                                return (0, 0, 0, pad_amount) # llama, falcon
+                            else:
+                                assert False, 'Unknown case, please handle, or dont use bucketing'
+                        new_kv = [None for i in range(len(model_kwargs['past_key_values']))]
+                        for i in range(len(model_kwargs['past_key_values'])):
+                            tmp_lst = [None for j in range(len(model_kwargs['past_key_values'][i]))]
+                            for j in range(len(model_kwargs['past_key_values'][i])):
+                                pad_tuple = create_pad_arg(pad_amount, i, j)
+                                assert model_kwargs['past_key_values'][i][j].shape[-(len(pad_tuple)//2)] % bucketsize == 0
```
Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1310]:
```suggestion
                                assert model_kwargs['past_key_values'][i][j].shape[-(len(pad_tuple)//2)] % bucket_size == 0
```

Comment (regisss) [inline optimum/habana/transformers/generation/utils.py:1310]: Why are we doing this assert? Maybe add a comment to explain a bit

File: examples/text-generation/run_generation.py [hunk 35]
```diff
@@ -535,19 +555,29 @@ def preprocess_function(examples):
         raw_dataset = raw_dataset.remove_columns([column_name])
         raw_dataset.set_format(type="torch")
 
-        separator = None
+        if prompt_length <= 0:
+            # Todo please check if this collate function is suitable for your model
+            # This has been tested for OPT, and Bloom
+            assert 'GaudiOPTForCausalLM' in str(type(model)) or 'GaudiBloomForCausalLM' in str(type(model))
```
Comment (regisss) [inline examples/text-generation/run_generation.py:561]:
```suggestion
            assert model.config.model_type in ["opt", "bloom"]
```

**Review by regisss — COMMENTED** [review]
I left many comments but they are mostly for changing `bucketsize` to `bucket_size` in order to stay consistent with other parameters like `batch_size`.

**Comment by ssarkar2** [comment]
@regisss @libinta :
Addressed regiss's comments

File: optimum/habana/transformers/generation/configuration_utils.py [hunk 36]
```diff
@@ -22,7 +22,11 @@ class GaudiGenerationConfig(GenerationConfig):
     limit_hpu_graphs (`bool`, *optional*):
         Skip HPU Graph usage for first token to save memory
     reuse_cache (`bool`, *optional*):
-         Whether to reuse key/value cache for decoding. It should save memory.
+        Whether to reuse key/value cache for decoding. It should save memory.
+    bucket_size (`bool`, *optional*):
```
Comment (regisss) [inline optimum/habana/transformers/generation/configuration_utils.py:26]:
```suggestion
    bucket_size (`int`, *optional*):
```

File: optimum/habana/transformers/generation/configuration_utils.py [hunk 37]
```diff
@@ -22,7 +22,11 @@ class GaudiGenerationConfig(GenerationConfig):
...
+        If negative (default=-1) pad to max if `static_shapes` is set. Else start with
+        `shape = bucket_size * ceil(prompt_len/bucket_size)` and then grow space by `bucket_size` when needed
```
Comment (regisss) [inline optimum/habana/transformers/generation/configuration_utils.py:28]:
```suggestion
        `shape = bucket_size * ceil(prompt_len/bucket_size)` and then grow space by `bucket_size` when needed.
```

File: optimum/habana/transformers/generation/configuration_utils.py [hunk 38]
```diff
@@ -22,7 +22,11 @@ class GaudiGenerationConfig(GenerationConfig):
...
+        Only active if `static_shapes` is used. Can't be used with `reuse_cache`
```
Comment (regisss) [inline optimum/habana/transformers/generation/configuration_utils.py:29]:
```suggestion
        Only active if `static_shapes` is used. Can't be used with `reuse_cache`.
```

**Review by regisss — COMMENTED** [review]
I just spotted a typo in the type of the docstring. I'm going to test it later today and we should be able to merge it by the end of the day!

**Review by regisss — APPROVED** [review]
LGTM!
I also pushed a change to modify one arg name in `run_generation.py`.
